File: greedy/leetcode_860/solution.py
from typing import List

class Solution:
    def lemonadeChange(self, bills: List[int]) -> bool:
        five_dollars = 0
        ten_dollars = 0

        for bill in bills:
            if bill == 5:
                five_dollars += 1

            elif bill == 10:
                five_dollars -= 1
                ten_dollars += 1

                if five_dollars < 0:
                    return False
            
            elif bill == 20:
                if five_dollars > 0 and ten_dollars > 0:
                    five_dollars -= 1
                    ten_dollars -= 1

                elif five_dollars >= 3:
                    five_dollars -= 3

                else:
                    return False

        return True 

File: greedy/leetcode_860/solution.js
/**
 * @param {number[]} bills
 * @return {boolean}
 */
var lemonadeChange = function(bills) {
    let fiveDollars = 0;
    let tenDollars = 0;

    for (const bill of bills ) {
        switch (bill) {
            case 5:
                fiveDollars ++;

                break;
            case 10:
                fiveDollars --;
                tenDollars ++;

                if(fiveDollars < 0) {
                    return false;
                }

                break;
            case 20:
                if (fiveDollars > 0 && tenDollars > 0) {
                    fiveDollars --;
                    tenDollars --;
                } else if(fiveDollars >= 3) {
                    fiveDollars -= 3;
                } else {
                    return false;
                }

                break;
        }
    }

    return true;
};